File: 课程/CaiPiao/CaiPiao/STBaseSettingViewController.m
//
//  STBaseSettingViewController.m
//  CaiPiao
//
//  Created by kemchenj on 5/22/16.
//  Copyright © 2016 kemchenj. All rights reserved.
//

#import "STBaseSettingViewController.h"
#import "STSettingCell.h"
#import "STGroupItem.h"
#import "STSettingItem.h"
#import "STArrowItem.h"



@implementation STBaseSettingViewController

- (void)viewDidLoad;
{
    [super viewDidLoad];
}

- (NSMutableArray *)groups;
{
    if (!_groups) {
        _groups = [NSMutableArray array];
    }
    
    NSLog(@"%@", _groups);

    return _groups;
}

- (instancetype)init;
{
    return [self initWithStyle:UITableViewStyleGrouped];
}



#pragma mark - <TableView DataSource>

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView;
{
    NSInteger i = self.groups.count;
    return i;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section;
{
    STGroupItem *group = _groups[section];
    return group.items.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath;
{
    STSettingCell *cell = [STSettingCell cellWithTableView:tableView
                                                     style:UITableViewCellStyleValue1
    ];

    STGroupItem *group = self.groups[indexPath.section];

    STSettingItem *settingItem = group.items[indexPath.row];

    cell.item = settingItem;

    return cell;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section;
{
    STGroupItem *item = self.groups[section];

    return item.headerTitle;
}

- (NSString *)tableView:(UITableView *)tableView titleForFooterInSection:(NSInteger)section;
{
    STGroupItem *item = self.groups[section];

    return item.footerTitle;
}



#pragma mark - <TableView Delegate>

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath;
{
    STGroupItem *group = self.groups[indexPath.section];

    STSettingItem *item = group.items[indexPath.row];

    if (item.operationBlock) {
        item.operationBlock(indexPath);

        return;
    }

    if ([item isKindOfClass:[STArrowItem class]]) {
        STArrowItem *arrowItem = (STArrowItem *) item;

        if (arrowItem.descVcClass == nil) {
            return;
        }

        UIViewController *controller = [[arrowItem.descVcClass alloc] init];

        [self.navigationController pushViewController:controller
                                             animated:YES
        ];
    }
}


@end

File: 课程/彩票/彩票/Setting(设置)/Controller/STBaseSettingViewController.swift
//
//  STBaseSettingViewController.swift
//  彩票
//
//  Created by kemchenj on 5/25/16.
//  Copyright © 2016 kemchenj. All rights reserved.
//

import Foundation
import UIKit



// MARK: - Class

class STBaseSettingViewController: UITableViewController {
    
    lazy var groups: NSMutableArray! = NSMutableArray()
    
}



// MARK: - Property

private extension STBaseSettingViewController {
    
    
    
}



// MARK: - UIView

extension STBaseSettingViewController {
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
}



// MARK: - TableView

extension STBaseSettingViewController {
    
    // MARK: + DataSource
    
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return groups.count
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let group = groups[section] as! STGroupItem
        
        return (group.items!.count)
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = STSettingCell.custom(tableView, style:.Value1) as! STSettingCell
        
        let group = self.groups[indexPath.section] as! STGroupItem
        
        cell.item = group.items![indexPath.row] as! STSettingItem
        
        return cell
    }
    
    override func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        let group = groups[section] as! STGroupItem
        
        return group.headerTitle
    }
    
    override func tableView(tableView: UITableView, titleForFooterInSection section: Int) -> String? {
        let group = groups[section] as! STGroupItem
        
        return group.footerTitle
    }
    
    
    // MARK: + Delegate
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let group = groups[indexPath.section] as! STGroupItem
        
        let item = group.items![indexPath.row] as! STSettingItem
        
        if let block = item.operationBlock {
            block(indexPath: indexPath)
            return
        }
        
        if item.isKindOfClass(STArrowItem) {
            let arrowItem = item as! STArrowItem
            
            if arrowItem.descVcClass == nil {
                return
            }
            
            let tempClass = arrowItem.descVcClass!.self
            
            let vc = tempClass()
            
            
        }
    }
}









